src/System.Private.CoreLib/src/System/Threading/WaitSubsystem.HandleManager.Unix.cs

```diff
@@ -0,0 +1,125 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Threading
+{
+    internal static partial class WaitSubsystem
+    {
+        [EagerStaticClassConstruction] // used during lazy class construction
+        private static class HandleManager
+        {
+            private const int InitialCapacity = 4;
+
+            /// <summary>
+            /// Prevents reusing a deleted index until at least there are this many deleted indexes in the free list. The
+            /// purpose is to allow free handles to remain free for some time so that we can discover invalid attempts at using
+            /// free handles.
+            /// </summary>
+            private const int MinimumFreeIndexCountBeforeReuse = 64;
+
+            private static WaitableObject[] s_objects = new WaitableObject[InitialCapacity];
+            private static int s_count;
+            private static LowLevelQueue<int> s_freeIndexes;
```

Comment:
What to have in a CoreLib and what to not have it is a fine balance. We did not always paid the right amount pf attention it.

Anyway, now that I slept on it I think the handle manager should be even more simpler: It should just allocate GC handles. Yes, you won't be able to detect some uses of invalid handles instead of crashing but I think it is absolutely fine.

It is very rare for managed code to operate on wait handles as IntPtrs. If there is such code and gets it wrong, it is fine for it to have different failure pattern. No need to be bug-for-bug compatible.

If we find that GCHandle lifetime is a common problem, we should build a debugging aid for it separately. It will cover this as well as all other GCHandle use cases.